### PR TITLE
feat(ingest): extract powerbi endorsements to tags

### PR DESCRIPTION
**Pull request by looppi**
Extract endorsements from workspace scan result to datasets, reports or dashboards.

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/docs/sources/powerbi/powerbi_recipe.yml [hunk 1]
```diff
@@ -16,6 +16,8 @@ source:
     client_secret: bar
     # Enable / Disable ingestion of user information for dashboards
     extract_ownership: true
+    # Enable / Disable ingestion of endorsements
+    extract_endorsements_to_tags: false
```
Comment (mohdsiddique) [inline metadata-ingestion/docs/sources/powerbi/powerbi_recipe.yml:20]: we can keep it true by default

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 2]
```diff
@@ -633,57 +673,6 @@ def get_pages_by_report(
             for raw_instance in response_dict["value"]
         ]
 
-    def get_reports(
```
Comment (mohdsiddique) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:636]: Any specific reason to delete this method?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py [hunk 3]
```diff
@@ -422,6 +443,16 @@ def chart_custom_properties(dashboard: PowerBiAPI.Dashboard) -> dict:
         if owner_mcp is not None:
             list_of_mcps.append(owner_mcp)
 
+        if self.__config.extract_endorsements_to_tags and dashboard.tags:
```
Comment (mohdsiddique) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:446]: Could we go for the append method? where append the MCP if flag is enabled to avoid checking at multiple places

**Review by mohdsiddique — CHANGES_REQUESTED** [review]

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py (hunk 2 above)
Comment (looppi) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:636]: It wasn't called anymore as the reports were parsed from the workspace scan. No other reason, basically the logic moved to the part where the scan result is handled.

The previous implementation was gathering the reports through the REST API, the currently the API doesn't include endorsementDetails in the payload, only the workspace scan has that data. That's the main reason why I ended up rewriting the logic.

Highly frustrating to have some data in the workspace scan and some data in the REST API.

File: metadata-ingestion/docs/sources/powerbi/powerbi_recipe.yml (hunk 1 above)
Comment (looppi) [reply metadata-ingestion/docs/sources/powerbi/powerbi_recipe.yml:20]: Sure!

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py (hunk 3 above)
Comment (looppi) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:446]: Meaning something like the `tags` in the dataclasses are populated only when `extract_endorsements_to_tags=True`? 

Then the code would look something like this, right?
`
for tag in dashboard.tags:
    list_of_mcps.append(
        self.new_mcp(
            Constant.DASHBOARD,
            dashboard_urn,
            Constant.GLOBAL_TAGS,
            self.transform_tags(dashboard.tags)
            tags,
        )
    )
`

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/config.py [hunk 4]
```diff
@@ -139,6 +140,10 @@ class PowerBiAPIConfig(EnvBasedSourceConfigBase):
     extract_lineage: bool = pydantic.Field(
         default=True, description="Whether lineage should be ingested"
     )
+    # Enable/Disable extracting endorsements to tags
+    extract_endorsements_to_tags: bool = pydantic.Field(
+        default=True, description="Whether to extract endorsements to tags"
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/config.py:145]: Can we default this to "False"? Most people may be surprised by this behavior, since this can overwrite existing Tags defined in DataHub

Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/config.py:145]: Please warn the author that enabling this will potentially overwrite Tags defined inside the DataHub application

Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/config.py:145]: You can also mention that there's an AddTagTransformer that supports semantic addition of tags (e.g. PATCH instead of REPLACE)

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 5]
```diff
@@ -374,34 +374,73 @@ def get_dashboards(self, workspace: Workspace) -> List[Dashboard]:
...
+        """
+        Get the list of dashboard from PowerBi for the given workspace identifier
+
+        TODO: Pagination. As per REST API doc (https://docs.microsoft.com/en-us/rest/api/power-bi/dashboards/get
+        -dashboards), there is no information available on pagination
+        """
+        dashboards: List[PowerBiAPI.Dashboard] = []
+        dashboard_data = self.get_dashboard_data(workspace)
+
+        for scanned_dashboard in scan_result["dashboards"]:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:398]: Please extract into a separate method. This is getting too long

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 6]
```diff
@@ -374,34 +374,73 @@ def get_dashboards(self, workspace: Workspace) -> List[Dashboard]:
...
+            # Iterate through response and create a list of PowerBiAPI.Dashboard
+            dashboard_details = next(
+                (x for x in dashboard_data if x["id"] == scanned_dashboard["id"]), None
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:401]: Is ID guaranteed to be present? Should would be defensive on this check?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 7]
```diff
@@ -374,34 +374,73 @@ def get_dashboards(self, workspace: Workspace) -> List[Dashboard]:
...
+            )
+            if not dashboard_details:
+                # Dashboard details was not found from REST API payload,
+                # which probably means we encountered an [App] Dashboard.
+                # Skip for now.
+                continue
+            # Iterate through response and create a list of PowerBiAPI.Dashboard
+            dashboard = PowerBiAPI.Dashboard(
+                id=scanned_dashboard.get("id"),
+                isReadOnly=dashboard_details.get("isReadOnly"),
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:411]: isReadOnly guaranteed to be there? Always?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 8]
```diff
@@ -374,34 +374,73 @@ def get_dashboards(self, workspace: Workspace) -> List[Dashboard]:
...
+                displayName=dashboard_details.get("displayName"),
+                embedUrl=dashboard_details.get("embedUrl"),
+                webUrl=dashboard_details.get("webUrl"),
                 workspace_id=workspace.id,
                 workspace_name=workspace.name,
                 tiles=[],
                 users=[],
+                tags=self.parse_endorsement(
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:419]: Is this a safe "get"? Will endorsementDetails always be there?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 9]
```diff
@@ -877,6 +868,45 @@ def init_dashboard_tiles(workspace: PowerBiAPI.Workspace) -> None:
 
             return None
 
+        def handle_report(report_data: dict) -> Optional[PowerBiAPI.Report]:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:871]: Why is this change required?

**Review by jjoyce0510 — COMMENTED** [review]
Overall, I'd recommend reducing the surface area of this PR. It changes too many things at once :) Any way to break it into a few smaller chunks?

Cheers
John

File: metadata-ingestion/docs/sources/powerbi/powerbi_recipe.yml (hunk 1 above)
Comment (looppi) [reply metadata-ingestion/docs/sources/powerbi/powerbi_recipe.yml:21]: As per John's comment, I'll keep this disabled by default.

**Comment by looppi** [comment]
I've refactored the implementation a bit and brought back the original implementation to handle fetching of reports. Now the implementation collects endorsements by id in a dict to the `Workspace` dataclass and those dicts are used when parsing dashboards and reports. As the datasets are parsed from the scan result directly, there was no need to implement that kind of handling, rather than use the scan result data directly.

With this change, there's no need to separate `[App]` reports and dashboards from the scan result and it "stays true" to the original implementation.

What do you think?

File: metadata-ingestion/docs/sources/powerbi/powerbi_recipe.yml (hunk 1 above)
Comment (mohdsiddique) [reply metadata-ingestion/docs/sources/powerbi/powerbi_recipe.yml:21]: Agree

File: metadata-ingestion/docs/sources/powerbi/powerbi_pre.md [hunk 10]
```diff
@@ -95,3 +95,9 @@ combine_result
 `Pattern-2` is *not* supported for upstream table lineage extraction as it uses nested item-selector i.e. {Source{[Schema="public",Item="book"]}[Data], Source{[Schema="public",Item="issue_history"]}[Data]} as argument to M-QUery table function i.e. Table.Combine
 
 `Pattern-1` is supported as it first assign the table from schema to variable and then variable is used in M-Query Table function i.e. Table.Combine
+
+## Extract endorsements to tags
+
+By default, extracting endorsement information to tags is disabled. The feature may be useful if organization uses [endorsements](https://learn.microsoft.com/en-us/power-bi/collaborate-share/service-endorse-content) to identify content quality.
+
+Please note that the default implementation overwrites tags for the ingested entities, if you need to preserve existing tags, consider using a [transformer](../../../../metadata-ingestion/docs/transformer/dataset_transformer.md#simple-add-dataset-globaltags) with `semantics: PATCH` tags instead of `OVERWRITE`.
```
Comment (jjoyce0510) [inline metadata-ingestion/docs/sources/powerbi/powerbi_pre.md:103]: Thank you!

**Review by jjoyce0510 — APPROVED** [review]
Changes look great!

Going to give this a big old "LGTM". Thanks for the hard work!
